### PR TITLE
CDK migration phase 2

### PR DESCRIPTION
**Pull request by TomPretty**
## What are you doing in this PR?

In this phase we use cdk to create a 'dual stack' of all of our infrastructure. We retain the `CfnInclude` construct to directly import our original cloudformation template but also recreate all of our resources directly with cdk (mainly with the GuEc2App construct). This means we will have a second load balancer and autoscaling group etc. We can deploy this to production and then, when we're happy, we can point fastly at the new load balancer to send traffic to the new stack. In a follow up PR, we can drop `CfnInclude` and delete our old cloudformation template.

To deploy this we need to:

1. Manually deploy the cloudformation change (it adds new parameters so riffraff will fail)
2. Merge PR and wait for build and deploy
3. Point fastly at the new load balancer

There are a number of `TODO` comments in the PR that are a result of DevEx highlighting certain areas of our infrastructure. These can be addressed as follow ups. It's safest to recreate our existing infrastructure exactly with cdk rather than trying to fix up any issues at the same time.

### CODE REVIEW

File: cdk/lib/frontend.ts
Comment (TomPretty) [inline cdk/lib/frontend.ts:50]: DevEx pointed out that these tables aren't region specific and that it might be preferable if they were

File: cdk/lib/frontend.ts
Comment (TomPretty) [inline cdk/lib/frontend.ts:73]: The `GuEc2App` construct will automatically give us permissions to `CODE/support/frontend` but the library that we use for config stores config in `support/frontend/CODE`

File: cdk/lib/frontend.ts
Comment (TomPretty) [inline cdk/lib/frontend.ts:171]: We only need this tag whilst we have a dual stack

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:96]: (This would save money)

File: support-frontend/conf/riff-raff.yaml
Comment (TomPretty) [inline support-frontend/conf/riff-raff.yaml:26]: We only need this whilst we're migrating

**Comment by TomPretty** [comment]
[support-frontend cdk migration - phase 2](https://trello.com/c/NnQsFrKC/358-support-frontend-cdk-migration-phase-2)

File: cdk/lib/frontend.ts
Comment (akash1810) [reply cdk/lib/frontend.ts:73]: Out of interest, what library are you using?

File: cdk/lib/frontend.ts
Comment (akash1810) [inline cdk/lib/frontend.ts:199]: Would be good to get this from parameter store (see also https://github.com/guardian/cdk/pull/884).

**Review by akash1810 — COMMENTED** [review]
The `GuAlarm`s being created are fairly similar - could we make their creation more terse by use of a for loop?

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:210]: Yeah makes sense - would this be a follow up to do after that cdk PR has been merged?

**Comment by TomPretty** [comment]
> The `GuAlarm`s being created are fairly similar - could we make their creation more terse by use of a for loop?

I had a go at removing some duplication. Mainly that was factoring out the repeated

```ts
this.withStageDependentValue({
      app,
      variableName: "AlarmEnabledInProd",
      stageValues: {
        [Stage.CODE]: false,
        [Stage.PROD]: true,
      },
    });
```

into a single `const` that can be reused. 

I'm not sure if we can squeeze it down much more given the other fields vary between alarms - unless you had something specific in mind?

File: cdk/lib/frontend.ts
Comment (i-hardy) [reply cdk/lib/frontend.ts:96]: I would be strongly in favour of this; I find it really hard to pick out details with Cloudwatch logs whereas Kibana has a much nicer UI.

**Comment by i-hardy** [comment]
I feel like I don't have enough familiarity with the existing setup to be the one to sign off on this, but I really, really love how much simpler and more understandable this is 🙏

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:150]: At the moment we are opting out of the 'standard' alarms that the pattern wants to create; we might be able to use some alarms provided via the pattern to reduce the number of custom alarms defined here.

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:238]: FYI - the pattern provides a helper for creating a high 5XX alarm. There is an example of minimal configuration for that [here](https://github.com/guardian/cdk/blob/d1de578b5691d83afa591a5b6c47cb91349ea70d/src/patterns/ec2-app/base.ts#L65-L75). The custom descriptions look important here, but I think you should be able to pass that in via the relevant alarm's [props](https://guardian.github.io/cdk/interfaces/constructs_cloudwatch.Http5xxAlarmProps.html).

At the moment, the pattern uses a % based alarm and combines two of your alarms (`High5XXRateAlarm` and `HighELB5XXRateAlarm`) into a single alarm. If using absolute thresholds and retaining the granularity of two alarms is important for your team, this might not be helpful for you.

The advantages of using the alarm provided via the pattern would be:

* reduction of boilerplate in this stack definition
* you will pick up improvements 'for free' if/when we establish best practices for monitoring/alerting in this area
* you can standardise this alarm across all of your apps more easily

However, as this is not a like for like replacement for you, it might be something to consider in a future PR.

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:285]: Interesting to see that you are alerting on latency; I wonder if we should provide this as standard @nicl @akash1810 @michaelwmcnamara?

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:213]: IIUC there is some overlap between the goal of this alarm and the goal of the [`GuUnhealthyInstances`](https://guardian.github.io/cdk/classes/constructs_cloudwatch.GuUnhealthyInstancesAlarm.html) alarm that the pattern is [able to create for you](https://github.com/guardian/cdk/blob/d1de578b5691d83afa591a5b6c47cb91349ea70d/src/patterns/ec2-app/base.ts#L73). It'd be interesting to know if the latter could work for you, but I think it makes sense to retain your existing alarm for now.

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:222]: I presume that this threshold is set to 2 because you want to run [at least 3 instances at all times in `PROD`](https://github.com/guardian/support-frontend/blob/c3c8289e84a301a72864ddf55cbddaa765594b48/cdk/lib/frontend.ts#L158). Now that you're defining all of this in TS, you could encode that logic (i.e. threshold = `minimumProdInstances - 1`), which means that this will continue working as desired if you ever scale up.

File: cdk/lib/frontend.ts
Comment (jacobwinch) [inline cdk/lib/frontend.ts:190]: I wonder if you really need a dedicated alarm for this; I'd have thought that `HighELB5XXRateAlarm` would also be triggered under this scenario (as long as you have a small amount of traffic).

File: cdk/lib/frontend.ts
Comment (jacobwinch) [reply cdk/lib/frontend.ts:73]: I think that [this code](https://github.com/guardian/support-frontend/blob/main/support-frontend/app/wiring/AppLoader.scala#L33) is [overriding](https://github.com/guardian/simple-configuration#custom-ssm-location) the library's [default path](https://github.com/guardian/simple-configuration#configurationloaderload).

Comment (TomPretty) [reply cdk/lib/frontend.ts:73]: Ahh so it looks like the default path from `simple-configuration` does align with what's generated for us

Comment (TomPretty) [reply cdk/lib/frontend.ts:73]: I'll create a health card to investigate this and see if we can clean it up (I'm guessing we can!)

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:238]: I think this would be a great improvement for all of those reasons that you stated. As it's not like for like I'll add a health card to address as a follow up so as to not block this PR going out

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:213]: I'll create a health card for this one too. I think it'd be great if we can fallback on as many defaults as possible from the `GuEc2App` pattern. I'm not convinced we need two separate alarms (`NoHealthyInstances` and `ReducedHealthInstances`), so if we can make this one work it would save quite a lot of boilerplate

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:190]: Yeah makes sense, again I've added something to chase this one up

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts]: Great suggestion! I've pushed that up

File: cdk/lib/frontend.ts
Comment (rupertbates) [reply cdk/lib/frontend.ts:50]: By this do you mean tagged to a specific AWS region?

File: cdk/lib/frontend.ts
Comment (rupertbates) [reply cdk/lib/frontend.ts:96]: The issue has always been around PII going into the central ELK stack, we'd need to check where we are with this currently

File: cdk/lib/frontend.ts
Comment (rupertbates) [reply cdk/lib/frontend.ts:212]: What is the advantage of getting it from parameter store? Wouldn't it just add another level of indirection making it harder to maintain and debug?

File: cdk/lib/frontend.ts
Comment (rupertbates) [reply cdk/lib/frontend.ts:311]: That would explain why we didn't get an alarm when the catalog was broken recently. I think we should probably fix this as it will bring the whole of subs down if we can't load the catalog.

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:50]: Yeah that's right. I would have to defer to DevX for exactly the benefits that would give us. I think it was that keeping everything region specific would isolate our infrastructure to just a single region. This would e.g make it easier to run another version of the app in another region (although I doubt we'd ever do it?). I think it might not give us much (any?) benefit here. I think it's more that 'keep things region specific' is on the DevX best practice checklist so they highlighted it as we went along

**Review by rupertbates — APPROVED** [review]
👍 nice one!

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:96]: That's good to know - I've created a [health card](https://trello.com/c/ARL9jV8g/367-send-support-frontend-logs-to-central-elk-stack) to investigate this, and specifically highlighted the issue around PII.

File: cdk/lib/frontend.ts
Comment (TomPretty) [reply cdk/lib/frontend.ts:311]: I've created a health card for this: https://trello.com/c/vUo5sO3R/368-fix-the-catalog-loading-failed-alarm-in-support-frontend

File: cdk/lib/frontend.ts
Comment (nicl) [reply cdk/lib/frontend.ts:290]: Yeah, this makes sense as latency is a form of availability. Frontend do something similar:

https://github.com/guardian/platform/blob/main/provisioning/cloudformation/frontend-service.yaml#L401

**Pull request merged**
